### ISSUE TITLE
TVICameraSource device sometimes nil

### ISSUE THREAD

**Issue by mcorner** [twilio/twilio-video-ios#37]
### Description

We are sporadically getting a nil device from our TVICameraSource.  This happens when we are trying to flip from the front camera to back.  We can reproduce it locally, but it only happens sometimes (<20% and usually on a first call).

### Steps to Reproduce
See code

#### Code

```
- (void)startPreview {
  // TVICameraSource is not supported with the Simulator.
  if ([self isSimulator]) {
    [self.previewView removeFromSuperview];
    return;
  }

  AVCaptureDevice *frontCamera = [TVICameraSource captureDeviceForPosition:AVCaptureDevicePositionFront];
  AVCaptureDevice *backCamera = [TVICameraSource captureDeviceForPosition:AVCaptureDevicePositionBack];

  self.camera = [[TVICameraSource alloc] initWithDelegate:self];
  self.localVideoTrack = [TVILocalVideoTrack trackWithSource:self.camera
  enabled:YES
  name:@"Camera"];
  // Add renderer to video track for local preview
  [self.localVideoTrack addRenderer:self.previewView];
  [self logMessage:@"Video track created"];

  if (frontCamera != nil || backCamera != nil) {
    [self.camera startCaptureWithDevice:frontCamera != nil ? frontCamera : backCamera completion:^(AVCaptureDevice *device, TVIVideoFormat *format, NSError *error) {
      if (error != nil) {
        [self log:[NSString stringWithFormat:@"Start capture failed with error.\ncode = %lu error = %@", error.code, error.localizedDescription]];
      } else {
        self.previewView.mirror = (device.position == AVCaptureDevicePositionFront);
      }
    }];
  } else {
    [self log:@"No front or back capture device found!"];
  }
}

RCT_EXPORT_METHOD(flipCamera) {
  if (!self.camera){
    [self log:@"Flipping: self.camera was nil!"];
    return;
  }

  if (!self.camera.device){
    [self log:@"Flipping: self.camera.device was nil!"];
    return;
  }

  AVCaptureDevice *newDevice = nil;

  if (self.camera.device.position == AVCaptureDevicePositionFront) {
    newDevice = [TVICameraSource captureDeviceForPosition:AVCaptureDevicePositionBack];
    if (!newDevice){
      [self log:@"Flipping: tried to flip to back, but new device is nil!"];
      return;
    }
  } else {
    newDevice = [TVICameraSource captureDeviceForPosition:AVCaptureDevicePositionFront];
    if (!newDevice){
      [self log:@"Flipping: tried to flip to front, but new device is nil!"];
      return;
    }
  }

  [self.camera selectCaptureDevice:newDevice completion:^(AVCaptureDevice *device, TVIVideoFormat *format, NSError *error) {
    if (error != nil) {
      [self log:[NSString stringWithFormat:@"Error selecting capture device.\ncode = %lu error = %@", error.code, error.localizedDescription]];
    } else {
      self.previewView.mirror = (device.position == AVCaptureDevicePositionFront);
    }
  }];
}```

#### Expected Behavior

Not be nil



#### Actual Behavior

This check fails:
  if (!self.camera.device){

#### Reproduces How Often

<20%, but only on a first call or install.

#### Logs

See code.

### Versions

2.7.0

#### Video iOS SDK

2.7.0

#### Xcode

10

#### iOS Version

12

#### iOS Device

iPhone X

**Comment by ceaglest** [comment]
Hi,

This is an interesting one. My first thought is that it could be permissions related or that the first starting of the camera hadn’t completed yet leading to TVICameraSource.device being nil. I will dig into this further on Monday.

Thanks,
Chris

**Comment by mcorner** [comment]
@ceaglest I couldn't reproduce it on an iphone 8, but we can (20%) of the time on the iphone X.  I thought it might be a race condition in setting the camera device, but it appears to still be nil even 2 minutes into the call.  If we start a new call on that device it is then fine.

The self.camera is not nil, just self.camera.device.

This is flipping from the front to rear camera.

**Comment by ceaglest** [comment]
Hi @mcorner,

> I couldn't reproduce it on an iphone 8, but we can (20%) of the time on the iphone X.

Ok, I have an iPhone X to try to reproduce locally. Do you get a completion block being fired after calling `startCaptureWithDevice`?

Best,
Chris

**Comment by mcorner** [comment]
@ceaglest My apologies, I believe we have found the source of our problem.

We are actually calling startPreview from two different places: starting a call and starting the preview view.  This is a react native app so we had to build some things ourselves.  On faster devices a race condition manifests and startPreview runs twice.

We put in a synchronized block and we believe it is working now. I will close and open it again if there is a problem. There is a chance that this might also help with #36, but I will leave that open until we confirm.
